src/ring_theory/fractional_ideal.lean

```diff
@@ -0,0 +1,417 @@
+/-
+Copyright (c) 2020 Anne Baanen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anne Baanen
+
+Fractional ideals of an integral domain.
+-/
+import ring_theory.algebra_operations
+import ring_theory.localization
+
+/-!
+# Fractional ideals
+
+This file defines fractional ideals of an integral domain and proves basic facts about them.
+
+## Main definitions
+
+ * `is_fractional` defines which `R`-submodules of `localization R S` are fractional ideals
+ * `fractional_ideal R S` is the type of fractional ideals in `localization R S`
+ * `has_coe (ideal R) (fractional_ideal R S)` instance
+ * `comm_semiring (fractional_ideal R S)` instance:
+   the typical ideal operations generalized to fractional ideals
+ * `lattice (fractional_ideal R S)` instance
+ * `has_div (fractional_ideal R (non_zero_divisors R))` instance:
+   the ideal quotient `I / J` (typically written $I : J$, but a `:` operator cannot be defined)
+
+## Main statements
+
+  * `mul_left_mono` and `mul_right_mono` state that ideal multiplication is monotone
+  * `right_inverse_eq` states that `1 / I` is the inverse of `I` if one exists
+
+## Implementation notes
+
+Fractional ideals are considered equal when they contain the same elements,
+independent of the denominator `a : R` such that `a I ⊆ R`.
+Thus, we define `fractional_ideal` to be the subtype of the predicate `is_fractional`,
+instead of having `fractional_ideal` be a structure of which `a` is a field.
+
+Most definitions in this file specialize operations from submodules to fractional ideals,
+proving that the result of this operation is fractional if the input is fractional.
+Exceptions to this rule are defining `(+) := (⊔)` and `⊥ := 0`,
+in order to re-use their respective proof terms.
+We can still use `simp` to show `I.1 + J.1 = (I + J).1` and `⊥.1 = 0.1`.
+
+We don't assume that `localization R S` is a field until we need it to define ideal quotients.
+When this assumption is needed, we replace `S` with `non_zero_divisors R`,
+making `localization R (non_zero_divisors R) = fraction_ring R` into a field since `R` is a domain.
+
+## References
+
+  * https://en.wikipedia.org/wiki/Fractional_ideal
+
+## Tags
+
+fractional ideal, fractional ideals, invertible ideal
+-/
+
+open localization
+
+set_option class.instance_max_depth 75
+
+universes u v w
+
+namespace ring
+
+section defs
+
+variables (R : Type u) [integral_domain R] (S : set R) [is_submonoid S]
+
+/-- A submodule `I` is a fractional ideal if `a I ⊆ R` for some `a ≠ 0`. -/
+def is_fractional (I : submodule R (localization R S)) :=
```

Comment:
When it comes to applying this theory, this definition would benefit enormously from using a characteristic predicate for localisations. For example, it will now be hard (i.e. impossible) to talk about fractional ideals in `rat` wrt to `int`. Because `rat` is not of the form `localization int _`.
@101damnations is working on bringing this characteristic predicate to mathlib. I hope that these two PR will play well together.

Reply:
I expect that changing `localization ℤ (non_zero_divisors ℤ)` to, let's say a typeclass `has_localization ℤ ℚ`, shouldn't have too much impact on this PR, since most of it is only concerned with `is_integer` and its lemmas.